### PR TITLE
Fix vectorization with inner broadcast axes

### LINKED ISSUE #1021: scheduler branch codegen error on single reduction
## 🐛 Bug

We are not properly generating vectorized write to global memory, which causes missing output elements in a single outer-dim reduction kernel that was vectorized on iteration axis.

```
Inputs:
  T0_g[ iS120{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, iS119{blockDim.x}, iS122{( ceilDiv(i5, 1) )}, iS121{1}, iS117{2}, iS123{1} ], float
  T1_g[ iblockIdx.x130{( ceilDiv(( ceilDiv(( ceilDiv(( i9 * ( i11 * i15 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x129{blockDim.x}, iS132{( ceilDiv(i13, 1) )}, iUS131{1}, iS127{2}, iS133{1} ], float
Outputs:
  T4_g[ iblockIdx.x82{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x81{blockDim.x}, iUS83{1}, iV79{2}, bS30{1} ] produce_pos( 3), float

%kernel_math {
T5_l[ iblockIdx.x100{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x99{blockDim.x}, iS102{( ceilDiv(i5, 1) )}, iUS101{1}, iV97{2}, iS103{1} ] ca_pos( 4 )
   = T0_g[ iS120{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, iS119{blockDim.x}, iS122{( ceilDiv(i5, 1) )}, iS121{1}, iS117{2}, iS123{1} ];
T6_l[ iblockIdx.x110{( ceilDiv(( ceilDiv(( ceilDiv(( i9 * ( i11 * i15 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x109{blockDim.x}, iS112{( ceilDiv(i13, 1) )}, iUS111{1}, iV107{2}, iS113{1} ] ca_pos( 4 )
   = T1_g[ iblockIdx.x130{( ceilDiv(( ceilDiv(( ceilDiv(( i9 * ( i11 * i15 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x129{blockDim.x}, iS132{( ceilDiv(i13, 1) )}, iUS131{1}, iS127{2}, iS133{1} ];
T2_l[ iblockIdx.x90{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x89{blockDim.x}, iS92{( ceilDiv(i5, 1) )}, iUS91{1}, iS87{2}, iS93{1} ] ca_pos( 6 ) produce_pos( 4)
   = T5_l[ iblockIdx.x100{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x99{blockDim.x}, iS102{( ceilDiv(i5, 1) )}, iUS101{1}, iV97{2}, iS103{1} ] ca_pos( 4 )
   + T6_l[ iblockIdx.x110{( ceilDiv(( ceilDiv(( ceilDiv(( i9 * ( i11 * i15 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x109{blockDim.x}, iS112{( ceilDiv(i13, 1) )}, iUS111{1}, iV107{2}, iS113{1} ] ca_pos( 4 );
T8_l[ iblockIdx.x52{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x51{blockDim.x}, rS54{( ceilDiv(i5, 1) )}rf, iUS53{1}, iS49{2}, iS55{1}rf ] ca_pos( 2 ) produce_pos( 6) = reduction( T2_l[ iblockIdx.x90{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x89{blockDim.x}, iS92{( ceilDiv(i5, 1) )}, iUS91{1}, iS87{2}, iS93{1} ] ca_pos( 6 ) produce_pos( 4), op = add, initial value = double(0) )
T3_l[ iblockIdx.x65{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x64{blockDim.x}, iUS66{1}, iS62{2}, rS67{1} ] ca_pos( 4 ) produce_pos( 2) = reduction( T8_l[ iblockIdx.x52{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x51{blockDim.x}, rS54{( ceilDiv(i5, 1) )}rf, iUS53{1}, iS49{2}, iS55{1}rf ] ca_pos( 2 ) produce_pos( 6), op = add, initial value = double(0) )
T7_l[ iblockIdx.x74{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x73{blockDim.x}, iUS75{1}, iS71{2}, bS18{1} ] ca_pos( 3 ) produce_pos( 4) = broadcast( T3_l[ iblockIdx.x65{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x64{blockDim.x}, iUS66{1}, iS62{2}, rS67{1} ] ca_pos( 4 ) produce_pos( 2) )
T4_g[ iblockIdx.x82{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x81{blockDim.x}, iUS83{1}, iV79{2}, bS30{1} ] produce_pos( 3)
   = T7_l[ iblockIdx.x74{( ceilDiv(( ceilDiv(( ceilDiv(( i1 * ( i3 * i7 ) ), 2) ), blockDim.x) ), 1) )}, ithreadIdx.x73{blockDim.x}, iUS75{1}, iS71{2}, bS18{1} ] ca_pos( 3 ) produce_pos( 4);
}
```

```
__global__ void kernel1(Tensor<float, 4> T0, Tensor<float, 4> T1, Tensor<float, 4> T4) {
  DEFINE_MAGIC_ZERO
  float T8[((1 * 2) * 1)];
  constexpr nvfuser_index_t ki351 = 0;
  #pragma unroll
  for(nvfuser_index_t ki349 = 0; ki349 < 2; ++ki349) {
    constexpr nvfuser_index_t ki347 = 0;
    T8[((((ki351 + 0) * 2) + ki349) * 1) + ki347] = 0;
  }
  UPDATE_MAGIC_ZERO
  #pragma unroll 1
  for(nvfuser_index_t ki309 = 0; ki309 < (ceilDiv(T0.size[2], 1)); ++ki309) {
    if (((((ki309 * 1) + (1 - 1)) < T0.size[2]) && (((((((blockIdx.x * 1) + (1 - 1)) * blockDim.x) + threadIdx.x) * 2) + (2 - 1)) < (T0.size[0] * (T0.size[1] * T0.size[3]))))) {
      constexpr nvfuser_index_t ki311 = 0;
      float T6[(2 * 1)];
      constexpr nvfuser_index_t ki334 = 0;
      reinterpret_cast<Array<float, 2>*>(&T6[((ki334 + 0) * 2) + (0 + 0)])->set(0);
      UPDATE_MAGIC_ZERO
      constexpr nvfuser_index_t ki330 = 0;
      *reinterpret_cast<Array<float, 2>*>(&T6[((ki330 + 0) * 2) + (0 + 0)]) = *reinterpret_cast<Array<float, 2>*>(&T1[((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * T0.size[2]) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * T0.size[2])) + (((ki309 * 1) + ki330) * (1 * T0.size[3])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]);
      UPDATE_MAGIC_ZERO
      float T5[(2 * 1)];
      constexpr nvfuser_index_t ki340 = 0;
      reinterpret_cast<Array<float, 2>*>(&T5[((ki340 + 0) * 2) + (0 + 0)])->set(0);
      UPDATE_MAGIC_ZERO
      constexpr nvfuser_index_t ki326 = 0;
      *reinterpret_cast<Array<float, 2>*>(&T5[((ki326 + 0) * 2) + (0 + 0)]) = *reinterpret_cast<Array<float, 2>*>(&T0[((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * T0.size[2]) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * T0.size[2])) + (((ki309 * 1) + ki326) * (1 * T0.size[3])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]);
      UPDATE_MAGIC_ZERO
      #pragma unroll
      for(nvfuser_index_t ki313 = 0; ki313 < 2; ++ki313) {
        constexpr nvfuser_index_t ki315 = 0;
        float T2[1];
        T2[0]
          = T5[((ki315 + 0) * 2) + (ki313 + 0)]
          + T6[((ki315 + 0) * 2) + (ki313 + 0)];
        T8[((((ki311 + 0) * 2) + ki313) * 1) + ki315]
          = T8[((((ki311 + 0) * 2) + ki313) * 1) + ki315]
          + T2[0];
      }
      UPDATE_MAGIC_ZERO
    } else {
      constexpr nvfuser_index_t ki311 = 0;
      float T6[(2 * 1)];
      constexpr nvfuser_index_t ki334 = 0;
      reinterpret_cast<Array<float, 2>*>(&T6[((ki334 + 0) * 2) + (0 + 0)])->set(0);
      UPDATE_MAGIC_ZERO
      constexpr nvfuser_index_t ki330 = 0;
      if (((((ki309 * 1) + (1 - 1)) < T0.size[2]) && (((((((blockIdx.x * 1) + (1 - 1)) * blockDim.x) + threadIdx.x) * 2) + (2 - 1)) < (T0.size[0] * (T0.size[1] * T0.size[3]))))) {
        *reinterpret_cast<Array<float, 2>*>(&T6[((ki330 + 0) * 2) + (0 + 0)]) = *reinterpret_cast<Array<float, 2>*>(&T1[((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * T0.size[2]) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * T0.size[2])) + (((ki309 * 1) + ki330) * (1 * T0.size[3])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]);
      }
      UPDATE_MAGIC_ZERO
      float T5[(2 * 1)];
      constexpr nvfuser_index_t ki340 = 0;
      reinterpret_cast<Array<float, 2>*>(&T5[((ki340 + 0) * 2) + (0 + 0)])->set(0);
      UPDATE_MAGIC_ZERO
      constexpr nvfuser_index_t ki326 = 0;
      if (((((ki309 * 1) + (1 - 1)) < T0.size[2]) && (((((((blockIdx.x * 1) + (1 - 1)) * blockDim.x) + threadIdx.x) * 2) + (2 - 1)) < (T0.size[0] * (T0.size[1] * T0.size[3]))))) {
        *reinterpret_cast<Array<float, 2>*>(&T5[((ki326 + 0) * 2) + (0 + 0)]) = *reinterpret_cast<Array<float, 2>*>(&T0[((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * T0.size[2]) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * T0.size[2])) + (((ki309 * 1) + ki326) * (1 * T0.size[3])) + (((((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]);
      }
      UPDATE_MAGIC_ZERO
      #pragma unroll
      for(nvfuser_index_t ki313 = 0; ki313 < 2; ++ki313) {
        constexpr nvfuser_index_t ki315 = 0;
        float T2[1];
        T2[0]
          = T5[((ki315 + 0) * 2) + (ki313 + 0)]
          + T6[((ki315 + 0) * 2) + (ki313 + 0)];
        if (((((ki309 * 1) + ki315) < T0.size[2]) && (((((((blockIdx.x * 1) + ki311) * blockDim.x) + threadIdx.x) * 2) + (ki313 + nvfuser_zero)) < (T0.size[0] * (T0.size[1] * T0.size[3]))))) {
          T8[((((ki311 + 0) * 2) + ki313) * 1) + ki315]
            = T8[((((ki311 + 0) * 2) + ki313) * 1) + ki315]
            + T2[0];
        }
      }
      UPDATE_MAGIC_ZERO
    }
  }
 if ((((((((blockIdx.x * 1) + (1 - 1)) * blockDim.x) + threadIdx.x) * 2) + (2 - 1)) < (T0.size[0] * (T0.size[1] * T0.size[3])))) {
    constexpr nvfuser_index_t ki291 = 0;
    float T7[2];
    #pragma unroll
    for(nvfuser_index_t ki299 = 0; ki299 < 2; ++ki299) {
      float T3[1];
      constexpr nvfuser_index_t ki305 = 0;
      T3[0]
         = T8[((((ki291 + 0) * 2) + ki299) * 1) + ki305];
      T7[(ki299 + 0)]
         = T3[0];
    }     
    UPDATE_MAGIC_ZERO
    T4[((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * 1) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * 1)) + (((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]
       = T7[(0 + 0)]; 
    UPDATE_MAGIC_ZERO 
  } else { 
    constexpr nvfuser_index_t ki291 = 0;
    float T7[2];
    #pragma unroll
    for(nvfuser_index_t ki299 = 0; ki299 < 2; ++ki299) {
      float T3[1];
      constexpr nvfuser_index_t ki305 = 0;
      T3[0]
         = T8[((((ki291 + 0) * 2) + ki299) * 1) + ki305];
      T7[(ki299 + 0)]
         = T3[0];
    }     
    UPDATE_MAGIC_ZERO
    if ((((((((blockIdx.x * 1) + (1 - 1)) * blockDim.x) + threadIdx.x) * 2) + (2 - 1)) < (T0.size[0] * (T0.size[1] * T0.size[3])))) {
      T4[((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) / (T0.size[1] * T0.size[3])) * (((1 * T0.size[3]) * 1) * T0.size[1])) + (((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) / T0.size[3]) * ((1 * T0.size[3]) * 1)) + (((((((((blockIdx.x * 1) + ki291) * blockDim.x) + threadIdx.x) * 2) + (0 + nvfuser_zero)) % (T0.size[1] * T0.size[3])) % T0.size[3]) * 1)]
         = T7[(0 + 0)]; 
    }
    UPDATE_MAGIC_ZERO
  }
}
```

Note that, `T8[2]` was computed on both elements and written to `T7[2]` correspondingly, but only `T7[0]` gets written to global memory `T4`

## To Reproduce

You will need to run this on a P100 or T4 card (so we'll have specs that triggers vectorization)
Repro branch: https://github.com/csarofeen/pytorch/tree/outer_reduction_vectorized_repro
Run it with `python test_jit_cuda_fuser.py TestCudaFuser.test_reduction_repro`

note. let's push fixes to this ongoing PR as well: https://github.com/csarofeen/pytorch/pull/979

### PR DESCRIPTION
**Pull request by naoyam**
Vectorization was disabled when broadcast inner axes exist.

Fixes #1021

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -15579,6 +15579,43 @@ TEST(NVFuserTest, FusionIssue1016_CUDA) {
...
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, new Double(1));
+  auto tv2 = broadcast(tv1, {false, true});
+  fusion.addOutput(tv2);
+
+  fusion.printMath();
```
Comment (rdspring1) [inline test/cpp/jit/test_gpu.cpp]: nit: remove debug prints

File: test/cpp/jit/test_gpu.cpp [hunk 2]
```diff
@@ -15579,6 +15579,43 @@ TEST(NVFuserTest, FusionIssue1016_CUDA) {
...
+  auto tv3 = tv2->cache_before();
+
+  tv2->split(0, 2);
+
+  tv1->computeAt(tv2, 1);
+
+  tv2->axis(0)->parallelize(ParallelType::TIDx);
+  tv2->axis(1)->parallelize(ParallelType::Vectorize);
+
+  fusion.printMath();
```
Comment (rdspring1) [inline test/cpp/jit/test_gpu.cpp]: nit: remove debug prints

**Review by rdspring1 — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
I'm looking at the ci failure. We have a script removed in upstream and some minor conflicts.
I'll take it from here. Thanks for stamping @rdspring1

**Comment by jjsjann123** [comment]
I'm verifying the fix, will merge after local tests.

**Pull request merged**